File: android/app/src/main/java/com/mochasmindlab/mlhealth/util/CsvParser.kt
package com.mochasmindlab.mlhealth.util

/**
 * A small RFC-4180-compliant CSV parser.
 *
 * Handles quoted fields, embedded commas, escaped quotes (""), and both LF and
 * CRLF line endings. Naive line-splitting breaks on food names that contain
 * commas, which is exactly what MyFitnessPal exports contain, so imports use
 * this instead. Mirrors the iOS CSVParser.
 */
object CsvParser {

    /** Parse raw CSV text into rows of string fields. */
    fun parse(text: String): List<List<String>> {
        var input = text
        // Strip UTF-8 BOM some exporters prepend.
        if (input.isNotEmpty() && input[0] == '﻿') input = input.substring(1)

        val rows = mutableListOf<List<String>>()
        val field = StringBuilder()
        var record = mutableListOf<String>()
        var inQuotes = false
        var i = 0

        fun endField() {
            record.add(field.toString())
            field.setLength(0)
        }
        fun endRecord() {
            endField()
            // Skip fully-empty records (a lone blank line).
            if (!(record.size == 1 && record[0].isEmpty())) {
                rows.add(record)
            }
            record = mutableListOf()
        }

        while (i < input.length) {
            val c = input[i]
            if (inQuotes) {
                if (c == '"') {
                    if (i + 1 < input.length && input[i + 1] == '"') {
                        field.append('"')   // escaped quote
                        i++
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(c)
                }
            } else {
                when (c) {
                    '"' -> inQuotes = true
                    ',' -> endField()
                    '\r' -> {
                        if (i + 1 < input.length && input[i + 1] == '\n') i++
                        endRecord()
                    }
                    '\n' -> endRecord()
                    else -> field.append(c)
                }
            }
            i++
        }
        if (field.isNotEmpty() || record.isNotEmpty()) endRecord()
        return rows
    }

    /**
     * Parse into a header list plus rows keyed by the (lower-cased, trimmed)
     * header. Missing trailing fields are treated as empty.
     */
    fun parseWithHeaders(text: String): Pair<List<String>, List<Map<String, String>>> {
        val raw = parse(text)
        if (raw.isEmpty()) return Pair(emptyList(), emptyList())
        val headers = raw.first().map { it.trim() }
        val rows = raw.drop(1).map { fields ->
            val map = mutableMapOf<String, String>()
            headers.forEachIndexed { idx, header ->
                map[header.lowercase()] = if (idx < fields.size) fields[idx] else ""
            }
            map
        }
        return Pair(headers, rows)
    }
}

File: ios/HealthTracker/Utils/CSVParser.swift
import Foundation

/// A small RFC-4180-compliant CSV parser.
///
/// Handles quoted fields, embedded commas, escaped quotes (""), and both
/// LF and CRLF line endings. The naive line-splitting used elsewhere
/// (e.g. USDACSVImporter) breaks on food names that contain commas, which is
/// exactly what MyFitnessPal exports contain, so imports use this instead.
enum CSVParser {

    /// Parse raw CSV text into rows of string fields.
    /// Empty trailing lines are ignored. Does not trim field whitespace
    /// (callers decide), except it strips a leading UTF-8 BOM if present.
    static func parse(_ text: String) -> [[String]] {
        var input = Substring(text)
        // Strip UTF-8 BOM some exporters prepend.
        if input.first == "\u{FEFF}" { input = input.dropFirst() }

        var rows: [[String]] = []
        var field = ""
        var record: [String] = []
        var inQuotes = false
        var i = input.startIndex

        func endField() {
            record.append(field)
            field = ""
        }
        func endRecord() {
            endField()
            // Skip fully-empty records (a lone blank line).
            if !(record.count == 1 && record[0].isEmpty) {
                rows.append(record)
            }
            record = []
        }

        while i < input.endIndex {
            let c = input[i]
            if inQuotes {
                if c == "\"" {
                    let next = input.index(after: i)
                    if next < input.endIndex && input[next] == "\"" {
                        field.append("\"")   // escaped quote
                        i = next
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(c)
                }
            } else {
                switch c {
                case "\"":
                    inQuotes = true
                case ",":
                    endField()
                case "\r":
                    // Handle CRLF: consume the following LF if present.
                    let next = input.index(after: i)
                    if next < input.endIndex && input[next] == "\n" { i = next }
                    endRecord()
                case "\n":
                    endRecord()
                default:
                    field.append(c)
                }
            }
            i = input.index(after: i)
        }
        // Flush the final field/record if the file didn't end with a newline.
        if !field.isEmpty || !record.isEmpty {
            endRecord()
        }
        return rows
    }

    /// Parse into dictionaries keyed by the (case-insensitive, trimmed) header row.
    /// Returns the header list and the rows. Missing trailing fields are treated as empty.
    static func parseWithHeaders(_ text: String) -> (headers: [String], rows: [[String: String]]) {
        let raw = parse(text)
        guard let headerRow = raw.first else { return ([], []) }
        let headers = headerRow.map { $0.trimmingCharacters(in: .whitespaces) }
        let rows: [[String: String]] = raw.dropFirst().map { fields in
            var dict: [String: String] = [:]
            for (idx, header) in headers.enumerated() {
                let key = header.lowercased()
                dict[key] = idx < fields.count ? fields[idx] : ""
            }
            return dict
        }
        return (headers, rows)
    }
}
